internal/signerverifier/dsse/dsse.go

```diff
@@ -44,15 +44,30 @@ func SignEnvelope(ctx context.Context, envelope *dsse.Envelope, signer dsse.Sign
 	}
 
 	pae := dsse.PAE(envelope.PayloadType, payload)
-	signature, err := signer.Sign(ctx, pae)
+	sigBytes, err := signer.Sign(ctx, pae)
 	if err != nil {
 		return nil, err
 	}
 
-	envelope.Signatures = append(envelope.Signatures, dsse.Signature{
-		Sig:   base64.StdEncoding.EncodeToString(signature),
+	signature := dsse.Signature{
+		Sig:   base64.StdEncoding.EncodeToString(sigBytes),
 		KeyID: keyID,
-	})
+	}
+
+	signatureIndex := -1
+	for i, sig := range envelope.Signatures {
+		if sig.KeyID == keyID {
+			signatureIndex = i
+			break
+		}
+	}
+
+	if signatureIndex > -1 {
+		// Replace existing signature for the key
+		envelope.Signatures[signatureIndex] = signature
+	} else {
+		envelope.Signatures = append(envelope.Signatures, signature)
+	}
```

Comment:
Why do we need this? Is it because the envelope passed in might already have signatures attached?

At minimum we should document the assumptions we have about the envelope signatures, but I also wonder if we should bother with this and leave the existing signatures 🤔 

If you do want to enforce the exactly 1 behavior of key ids <-> signatures, then this isn't sufficient since this only updates the first instance of a key id, but if the key id appears multiple times in the existing envelope the rest remain unmodified. 🤷

Reply:
I've documented in the docstring the correct behaviour. I've also updated this to scan the full list of signatures, keeping signatures from other keys and appending the new sig from the current key. This removes duplicates when they already exist in an envelope, and prevents duplicates in gittuf signed metadata (assuming key ID calculation is consistent).